design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,280 @@
+| 5  | ProviderDiscoveryTimeout (couldn't make connection with provider-uri) | Azure Identity Provider failed with timeout error (Provider URI: '{0}'). Reason: '{1}' |                                                                                      |        | Yes (needed to be generalized)                |
+| 6  | ProviderTokenInvalid (failed to confirm token signature of provider)  | Failed to confirm signature of '{0-token}' issued by (Provider URI: '{1}'              |                                                                                      |        | No                                            |
+| 7  | MissingRequestParam                                                   | Field '{0-field-name}' is missing or empty in request body                             | "Token" not in body of request                                                       |        | Yes (must extract authn-request body)         |
+| 8  | TokenFieldNotFoundOrEmpty                                             | Field '{0-field-name}' not found or empty in token                                     |                                                                                      |        | No (for OIDC but need to be more generalized) |
+| 9  | InvalidApplicationIdentity                                            | Azure application identity field '{0-field-name}' does not match Azure token           | 1+ of the fields in the Azure token don't match the Conjur host identity annotations |        | No                                            |
+ 
+## Docs
+
+## Version update
+
+- Conjur
+
+- DAP
+
+## DoD
+
+- [ ] Solution design is pushed to GH and project-artifacts slack channel for early feedback
+- [ ] Solution designed is approved
+- [ ] Test plan is approved
+- [ ] Security review was done
+- [ ] Implement Azure authn functionality
+- [ ] Azure VM can authenticate to Conjur using the Conjur authn-azure authenticator and get Conjur access token
+- [ ] Tests are implemented according to test plan and all tests pass
+    - [ ] Integration tests exist
+    - [ ] Unit tests exist
+- [ ] Security action items were taken care of
+- [ ] Performance tests were done and align with SLA
+- [ ] Logs were reviewed by TW and PO
+- [ ] Documentation has been written and reviewed by PO and TW for both DAP and Conjur
+- [ ] Engineer(s) that were not involved in the project use documentation to authenticate with Conjur using Azure
+- [ ] Versions are bumped in all relevant projects
+   - [ ] Conjur
+   - [ ] DAP
+ 
+## Open questions
+
+- Do we want to develop a mechanism to support a request body with more than one field (currently, we have only "token", but we may want to add more fields later).
```

Comment:
1. What would happen if a user will try to use Azure authenticator?
2. Do we need to validate the source of the token or provider to be Azure? 
3. Do you know of or think on any reason (other than tests) that an application would not be able to run from Azure Functions or App Service or  VM scale set?

Reply:
1. @InbalZilberman i don't understand the question. Can you elaborate?
2. We validate the token in `validate_azure_token` as part of the decoding.
3. Not that i can think of.

Reply:
@orenbm 
1. What would happen if a user, not a host, that is recognized in Azure would use the Azure Authenticator?
2. Do we validate that the provider is an Azure one? can it also be Okta for example? Can it be GKE? (any Oauth 2.0 provider)
3. @shaharglazner  WDYT on the third question?

Reply:
1. So the admin defined everything correctly but instead of using `!host` he used `!user`? With all the annotations and permissions?
2. We do not validate such a thing. We do validate that the `xms_mirid` field is present and has the correct application identity so while the validation of the token will pass (as it's an Oauth 2.0 token) it will not pass Conjur's authentication process.

Reply:
1. I would imagine having !user instead of !host will not break anything because in conjur user and host and group and layer are interchangeable and are more constructs to assist the customer

Reply:
The general validation are ignorant to the role type (host/user). We can strive to do that also in Azure if there is a business need for that. I don't see any implementation blocker on this.

@sigalsax - please update this in the doc when you have the time. Don't forget about test plan changes. 
@InbalZilberman  ^

Reply:
Regarding 1 let's decide if this use case is in or out of scope.

Reply:
let's go for it. Should be an easy win.